rosbag2_cpp/include/rosbag2_cpp/rewriter.hpp

```diff
@@ -0,0 +1,103 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef ROSBAG2_CPP__REWRITER_HPP_
+#define ROSBAG2_CPP__REWRITER_HPP_
+
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "rosbag2_cpp/converter_options.hpp"
+#include "rosbag2_cpp/reader.hpp"
+#include "rosbag2_cpp/writer.hpp"
+
+#include "rosbag2_storage/serialized_bag_message.hpp"
+#include "rosbag2_storage/storage_options.hpp"
+
+// This is necessary because of using stl types here. It is completely safe, because
+// a) the member is not accessible from the outside
+// b) there are no inline functions.
+#ifdef _WIN32
+# pragma warning(push)
+# pragma warning(disable:4251)
+#endif
+
+namespace rosbag2_cpp
+{
+
+/**
+ * The Rewriter will open a bagfile and rewrite it into another bagfile.
+ * This can be used to "reshape" bagfiles, or even convert storage implementation,
+ * compression, or serialization format by passing in different options (storage or converter)
+ */
+class ROSBAG2_CPP_PUBLIC Rewriter final
+{
+public:
+  explicit Rewriter(
+    std::unique_ptr<Reader> reader_ =
+    std::make_unique<Reader>(),
+    std::unique_ptr<Writer> writer_ =
+    std::make_unique<Writer>());
+
+  ~Rewriter();
+
+  /**
+   * Opens an existing bagfile, and rewrites it into another bagfile
+   * with new settings to a new location.
+   *
+   * \param input_storage_options Storage options for the input bagfile
+   * \param input_converter_options Converter options for the input bagfile
+   * \param output_storage_options Storage options for the output bagfile
+   * \param output_converter_options Converter options for the output bagfile
+   **/
+  void rewrite(
+    rosbag2_storage::StorageOptions input_storage_options,
+    ConverterOptions input_converter_options,
+    rosbag2_storage::StorageOptions output_storage_options,
+    ConverterOptions output_converter_options
+  );
+
+  /**
+   * Opens a a vector of existing bagfiles, and rewrites it into another bagfile
+   * with new settings to a new location.
+   *
+   * \note the API consumer is responsible for sorting the input_storage_options according to desired order.
+   *
+   * \param input_storage_options Storage options for the input bagfile.
+   * \param input_converter_options Converter options for the input bagfile
+   * \param output_storage_options Storage options for the output bagfile
+   * \param output_converter_options Converter options for the output bagfile
+   **/
+  void rewrite_many(
```

Comment:
I think `rewrite` would be used for "conversion" (#717 , #599) but since the `rosbag2_cpp::Writer` cannot open a bag for rewriting , there had to be a way to only open the bag once for writing while opening multiple readers, and therefore I thought `rewrite_many` was appropriate.


(https://github.com/ros2/rosbag2/blob/cc12254cfd39d5c05918db19abd8d4d06ba007b2/rosbag2_cpp/include/rosbag2_cpp/writer.hpp#L65)

Reply:
Are you trying to combine multiple "bags", or multiple "files"? This remains unclear from this comment. A single Reader can read a multi-file bag. Can you clarify what you're trying to do here?

Reply:
Also, I'll note that "convert" is a heavily overloaded term that can refer to a variety of independent features, as we're learning in the conversation on #717, so I'd avoid the term right now in favor of more specific language.

Reply:
@emersonknapp I am sorry I am having a bit of trouble understand what you mean about "bags" v.s. "files".

Is there a minimal example you could point me to about reading multifiles with the reader? I found the test, but I am having trouble following
https://github.com/ros2/rosbag2/blob/master/rosbag2_cpp/test/rosbag2_cpp/test_multifile_reader.cpp

From my understanding, the reader takes in either a string URI or StorageOptions:
https://github.com/ros2/rosbag2/blob/5f77e8fbf13920c33a3ddf8f156d4840e9e82244/rosbag2_cpp/include/rosbag2_cpp/reader.hpp#L92
https://github.com/ros2/rosbag2/blob/5f77e8fbf13920c33a3ddf8f156d4840e9e82244/rosbag2_storage/include/rosbag2_storage/storage_options.hpp#L23

Reply:
In rosbag2, a "bag" is not a single file on disk, it's a directory, potentially containing many individual files.

For example if I do a recording like the following, observe the contents of the bag from `ls`

```
$ ros2 bag record -a -o mybag
[INFO] [1618420012.176030068] [rosbag2_storage]: Opened database 'mybag/mybag_0.db3' for READ_WRITE.
[INFO] [1618420012.176144426] [rosbag2_recorder]: Listening for topics...
[INFO] [1618420012.178912237] [rosbag2_recorder]: Subscribed to topic '/rosout'
[INFO] [1618420012.180516824] [rosbag2_recorder]: Subscribed to topic '/parameter_events'
[INFO] [1618420024.978740077] [rosbag2_recorder]: Subscribed to topic '/chatter'
^C[INFO] [1618420027.180706513] [rclcpp]: signal_handler(signal_value=2)
[INFO] [1618420027.387157257] [rosbag2_cpp]: Writing remaining messages from cache to the bag. It may take a while

$ ls mybag
metadata.yaml  mybag_0.db3
```

In this case, `mybag` is the "ROS 2 Bag" - and `mybag/mybag_0.db3` is a single _file_ within that bag.

You can see in a case with splitting - (`--max-bag-size` causes splitting when a file gets large enough) - that there are multiple files within the bag.

```
$ rm -rf mybag

$ ros2 bag record -a -o mybag --max-bag-size 86016
[INFO] [1618420123.290318343] [rosbag2_storage]: Opened database 'mybag/mybag_0.db3' for READ_WRITE.
[INFO] [1618420123.290484810] [rosbag2_recorder]: Listening for topics...
[INFO] [1618420123.292763350] [rosbag2_recorder]: Subscribed to topic '/rosout'
[INFO] [1618420123.294727032] [rosbag2_recorder]: Subscribed to topic '/parameter_events'
[INFO] [1618420123.296239889] [rosbag2_recorder]: Subscribed to topic '/chatter'
[INFO] [1618420127.495646435] [rosbag2_cpp]: Writing remaining messages from cache to the bag. It may take a while
[INFO] [1618420127.498400134] [rosbag2_storage]: Opened database 'mybag/mybag_1.db3' for READ_WRITE.
^C[INFO] [1618420129.424583789] [rclcpp]: signal_handler(signal_value=2)
[INFO] [1618420129.627688074] [rosbag2_cpp]: Writing remaining messages from cache to the bag. It may take a while

$ ls mybag
metadata.yaml  mybag_0.db3  mybag_1.db3
```

When you pass a `storage_uri` to `rosbag2` - you pass the _directory_ `mybag` - e.g. `ros2 bag play mybag`, you aren't passing it the path of the individual file. It then, using the metadata, knows to open the files in sequence automatically using the `SequentialReader`.

`StorageOptions` contains a `storage_uri` string, so it's a superset of configuration for the reader - both of them are expecting this directory.

Given this context - are you trying to "combine files within a split bag into a new single-file bag". Or are you trying to "combine multiple bags with their own metadata into a single bag"

Reply:
ah! ok I did not realize the `storage_uri` was describing the _directory_.  So essentially, I am trying to  "combine files within a split bag into a new single-file bag" , so I do not need the `rewrite_many` function yes.